src/__tests__/errorCalculation.js

```diff
@@ -17,7 +17,32 @@ describe('errorCalculation test', () => {
       data.y[i] = sampleFunction(i);
     }
 
-    expect(errorCalculation(data, [2, 2], sinFunction)).toBeCloseTo(0, 3);
-    expect(errorCalculation(data, [4, 4], sinFunction)).toBeCloseTo(48.7, 1);
+    expect(errCalc.sumOfResiduals(data, [2, 2], sinFunction)).toBeCloseTo(0, 5);
+    expect(errCalc.sumOfResiduals(data, [4, 4], sinFunction)).toBeCloseTo(48.7, 1);
+
+    expect(errCalc.sumOfSquaredResiduals(data, [2, 2], sinFunction)).toBeCloseTo(0, 5);
+    expect(errCalc.sumOfSquaredResiduals(data, [4, 4], sinFunction)).toBeCloseTo(165.6, 1);
+  });
+});
+
+describe('error propagation estimator', () => {
```

Comment:
@m93a Could you please help me understand these test values? When I look at them it is not at all obvious to me what the correct values should be. (Partly because I am not sure how `errorPropagation` works other than that it seems to take pseudo-random steps around a point and estimate the slope there)

Reply:
The `errorPropagation` method tries to approximate the [propagation of uncertainty](https://en.wikipedia.org/wiki/Propagation_of_uncertainty) through the function, given a variable with known error. Since we know the mean, as well as the standard deviation of the variable, we can assume it's normally distributed and choose equiprobable intervals where the variable's measured value might actually be.

This means that if we slice the interval of all variable's possible values to eg. 10 equiprobable intervals, there's the same probability that the measured value will be in the first interval, as there is probability that it will be in the second, third, etc. interval. For practical purposes, I don't really count with *all* of the possible values, I only use values between x̅ − 2.25σ and x̅ + 2.25σ. (There's 97.6 % probability that the measured value will be between those values). This process is not pseudo-random, but completely deterministic – the intervals are computed using the [CDF of normal distribution](https://en.wikipedia.org/wiki/Normal_distribution#Cumulative_distribution_function). 

Then, the functional values on the boundaries of the intervals are computed. This way the function can be approximated as a linear function on each interval. When a variable “passes through” a linear function, its standard error gets “stretched” (ie. multiplied) by the slope of the function. If a variable has the same probability of passing through multiple linear functions, its standard error (on average) gets stretched by the average of the slopes of those linear functions.

Of course the error distribution of the functional value most probably won't be a Gauss curve, even if the input variable was normally distributed in the first place. But as our fitting algorithm can't deal with those either (ie. it's not [robust](https://en.wikipedia.org/wiki/Robust_regression)), it's not a big deal. We can simply pretend that all of the resulting distributions are *close enough* to the normal distribution (which is mostly true as long as the errors are small).

(An uncanny fact as a bonus: if the error is large enough and the function is very unsymmetrical, the mean of the functional value won't be equal to the functional value of the mean. For now let's just pretend those cases never occur.)

Now about the test values. You're right that the values are not exactly obvious. The results should match the standard deviation of normally distributed data that passed through the function. One could use Monte Carlo to get the right value.

```javascript
// Pseudo-code to validate the results of errorPropagation
const mean = 42;
const sigma = 0.1;
let data = Array(BILLIONS_AND_BILLIONS)
data = data.map(() => normal.random(mean, sigma));
data = data.map( x => fn(x) );
const propagatedSigma = stdev(data);

expect(errCalc.errorPropagation(fn, mean, sigma))
  .toBeCloseTo(propagatedSigma, DEPENDS_ON_NUMBER_OF_INTERVALS);
```

Wow, this was really time-consuming, I'm taking a break for today. I've got other things to do 😉

Reply:
Thanks for the detailed explanation! I plan to read through it a few times and attempt to make the test cases more expressive once I understand it.